compiler/rustc_mir/src/const_eval/eval_queries.rs

```diff
@@ -304,6 +330,15 @@ pub fn eval_to_allocation_raw_provider<'tcx>(
     let res = ecx.load_mir(cid.instance.def, cid.promoted);
     match res.and_then(|body| eval_body_using_ecx(&mut ecx, cid, &body)) {
         Err(error) => {
+            if key.param_env.reveal() == Reveal::Selection {
+                match error.kind {
+                    err_inval!(Layout(LayoutError::Unknown(_)))
+                    | err_inval!(TooGeneric)
+                    | err_inval!(AlreadyReported(_)) => {}
```

Comment:
Why are those exact three treated differently here?

Reply:
> That will cause some new issues though -- it will be much harder to avoid printing the same const error multiple times.

when const eval prints an error the `InterpError` is `AlreadyReported`. So we don't return `Silent` here to prevent emitting the error twice.

`err_inval!(Layout(LayoutError::Unknown(_))) | err_inval!(TooGeneric)` because these result in `ErrorHandled::TooGeneric` which we still want to emit

Probably needs a comment as well ^^

Reply:
You are quoting me entirely out of context here.^^ In that quote I was referring to "have the query return a Result".

Reply:
> when const eval prints an error the InterpError is AlreadyReported. So we don't return Silent here to prevent emitting the error twice.

Well but it means we hit the code path below for reporting it? Or is `report_as_lint` et al silent for `AlreadyReported`?

> err_inval!(Layout(LayoutError::Unknown(_))) | err_inval!(TooGeneric) because these result in ErrorHandled::TooGeneric which we still want to emit

Oh we do? I thought TooGeneric should never be emitted, it means "please re-evaluate with `RevealAll`?